### PR TITLE
feat: Redesigns purchase screen

### LINKED ISSUE #9: Integrate @atb-as/theme
Add `@atb-as/theme` as a dependency and use it in the current CSS.

### LINKED ISSUE #22: Component: Button
- [ ] Design ready
- [ ] Styles: Primary, secondary
- [ ] States: Normal, focus, hover, active, disabled

### LINKED ISSUE #27: Component: Message
This is intended to be put inside a section, but could be put elsewhere as well.

- [ ] Design ready
- [ ] Styles: Info, Warning, Error, Valid

### LINKED ISSUE #28: Component: Section
A section is sometimes shown as an accordion, but not always. Some design/evaluation work is needed here.

Items in a section are:

- Sometimes presented as an interactive component such as a button, input fields or a toggle -- and sometimes just non-interactive information/content.
- Sometimes expandable, but not always.
- Sometimes in multiple columns, but not always.

### LINKED ISSUE #36: Update design for Buy Ticket screen.
Update buy ticket screen flow to the new design.

image

https://www.figma.com/file/1Px0XvoDe2JHb6hmViNqXj/Webshop-0.1?node-id=604%3A2052

### PR DESCRIPTION
**Pull request by mikaelbr**
Denne PR-en blir i det største laget så jeg splitter opp feature i flere tasks. Som en del av arbeidet blir det også lagt om til gjenbrukbare komponenter. Innfører de først og fremst under Shop-skjermen, men dette kan refaktoriseres over til andre sider etterhvert. For å kunne jobbe i parallell på disse komponentene burde denne PR-en merges tidlig ogsa.

## Dekt av denne PR-en

- Introduserer UI komponenter: Button, Message (status), Section, radio buttons, radio group, section, etc. Dette er grunnleggende byggesteiner spesifisert etter design system i Figma. Det er foreløpig ikke lagt opp noe eksempel side på disse i bruk. Det kan løses i en separat oppgave og prioriteres frittstående. Burde gå over til å bare bruke disse komponentene da de er lagt opp til å være universelt utformet.
- Koblet opp notifications til å vises i 10 sekunder som globale meldinger. Feilhåndtering burde skje nærme feil og gi mulighet til å korrigere opp, men som en nødløsning kan vi sende en global feilmelding. Dette typisk på ting som ikke burde skje. Nå har jeg brukt det på reservasjon og uthenting av pris. Dette er egentlig ting som aldri burde skje. Kan utvides senere til å gi statusmelding direkte på kostnadsoppsummeringsnivå.
- Endrer grunnleggende designrammer til å være responsivt. Kun tatt i bruk i Shop foreløpig.
- Inkludert pakke for inert polyfill (brukt i accordion komponenten) og hjelpefunksjoner Html extra for enklere håndtering av conditional HTML og attributter.

## Gjenstående oppgaver

- https://github.com/AtB-AS/webshop/issues/39
- https://github.com/AtB-AS/webshop/issues/40
- https://github.com/AtB-AS/webshop/issues/41

## Eksempel skjermbilde

Se kjørende løsning for demo, men her er skjermbilde som illustrasjon:
Screenshot 2021-04-15 at 12 21 19

Markerer disse som fikset nå, selv om det kan være flere ting som må gjøres. Men det får vi ta som en del av videreutvikling av andre oppgaver og etter behov.

Fixes #36 
Fixes #22
Fixes #27
Fixes #28
Fixes #9

### CODE REVIEW

File: src/elm/Ui/Input.elm [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import Html as H exposing (Attribute, Html)
+import Html.Attributes as A
+import Html.Events as E
+import Html.Extra exposing (viewMaybe)
+import Ui.Button exposing (ButtonMode(..))
+import Ui.Group
+import Ui.TextContainer as Text exposing (TextColor(..), TextContainer(..))
+
+
+type alias RadioOptions msg =
```
Comment (mjenssen) [inline src/elm/Ui/Input.elm:15]: This is quite big, have you thought about making this component into a builder pattern? So an `init` function that returns a default value, and then setters to change the fields, and then finally a `view` function. In use, you would end up with something like:

```elm
Radio.init
    |> Radio.setTitle "My radio"
    |> Radio.setChecked True
    |> Radio.setOnCheck myOnCheck
    |> Radio.view
```

The `init` function can of course take things that are _always_ needed or useful as input, e.g. in this case it might not make sense to not have a title.

Another huge benefit with the builder pattern is that you can expand this record with more fields without having to update the entire codebase.

File: src/static/styles/main.scss [hunk 2]
```diff
@@ -19,31 +18,28 @@ body {
...
-    text-align: center;
-    font-weight: 500;
-
-    .go-back {
-        position: absolute;
-        left: 24px;
-        top: 24px;
+input,
+button {
+    font-family: var(--font-main);
```
Comment (mjenssen) [inline src/static/styles/main.scss:27]: I see that the text in (text) input fields is tiny compared to what it was before, at least in Firefox. Is this the culprit?

File: src/elm/Main.elm [hunk 3]
```diff
@@ -11,7 +12,7 @@ import Html as H exposing (Html)
 import Html.Attributes as A
 import Json.Decode as Decode exposing (Decoder)
 import Json.Decode.Pipeline as DecodeP
-import Notification exposing (Notification)
+import Notification exposing (Notification, decrementTimer)
```
Comment (mjenssen) [inline src/elm/Main.elm:15]: Please only expose types. It's much easier to see where things are when they are fully qualified. Types usually have a relation to their containing module (and there are fewer of them) so it's much easier to see where they are coming from.

File: src/elm/Page/Shop.elm [hunk 4]
```diff
@@ -380,190 +397,159 @@ richActionButton active maybeAction content =
...
-            NotLoaded ->
-                H.text ""
 
-            Loading _ ->
-                H.p [] [ H.text "Reserving offers..." ]
+maybeBuyNotice : List ( UserType, Int ) -> Html msg
+maybeBuyNotice users =
+    let
+        reduced =
+            users |> List.any (Tuple.first >> hasReducedCost)
```
Comment (mjenssen) [inline src/elm/Page/Shop.elm:541]: Just put  `users` at the end and skip the pipe?

File: src/elm/Page/Shop.elm [hunk 5]
```diff
@@ -20,7 +21,10 @@ import Set
 import Shared exposing (Shared)
 import Task
 import Time
-import Ui.Message as Message
+import Ui.Button exposing (ThemeColor(..))
+import Ui.Group
+import Ui.Input as Input
+import Ui.Message as Message exposing (defaultOption)
```
Comment (mjenssen) [inline src/elm/Page/Shop.elm:27]: `defaultOption` doesn't seem to be used?

File: src/elm/Ui/Button.elm [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    , primaryDefault
+    , secondary
+    , secondaryDefault
+    , tertiary
+    )
+
+import Html as H exposing (Html)
+import Html.Attributes as A exposing (disabled)
+import Html.Events as E
+import Html.Extra as Html
```
Comment (mjenssen) [inline src/elm/Ui/Button.elm:16]: Isn't this a bit confusing? Wouldn't it be better to not use `as Html` and then use `Html.Extra.viewMaybe` further down? Or at least give it a better name like `HtmlExtra` or even `HE`.

File: src/elm/Page/Shop.elm [hunk 7]
```diff
@@ -4,13 +4,14 @@ import Base exposing (AppInfo)
 import Data.RefData exposing (FareProduct, LangString(..), TariffZone, UserProfile, UserType(..))
 import Data.Ticket exposing (Offer, PaymentStatus, PaymentType(..), Reservation)
 import Environment exposing (Environment)
-import Fragment.Button as Button
 import Fragment.Icon as Icon
 import GlobalActions as GA
 import Html as H exposing (Html)
 import Html.Attributes as A
 import Html.Events as E
+import Html.Extra exposing (viewMaybe)
```
Comment (mjenssen) [inline src/elm/Page/Shop.elm:12]: Please don't expose functions. 😄

**Review by mjenssen — CHANGES_REQUESTED** [review]
Looks pretty good, mostly some nits and possibly some patterns that would be nice to follow for the customizable components. I also think we should be more consistent with camelCase, kebab-case, and snake_case, especially in CSS.

File: src/static/styles/main.scss (hunk 2 above)
Comment (mikaelbr) [reply src/static/styles/main.scss:27]: Inputs should be explicitly set as 16px here I think. This is also to prevent iOS to zoom in on focus.

File: src/elm/Ui/Input.elm (hunk 1 above)
Comment (mikaelbr) [reply src/elm/Ui/Input.elm:15]: I was the builder pattern was used in Notification, but noticed it too late. I decided on using record types early on to prevent ambiguous order and easier expanding (also per Rich Hickey's teachings 😬 ). I can explore rewriting this

Comment (mjenssen) [reply src/elm/Ui/Input.elm:15]: Oh yeah, don't get me wrong. Using records over tuples or a bazillion function arguments is really good, it makes it much easier to deal with. Rewriting it isn't strictly needed in this PR, but something to think about.

**Comment by mikaelbr** [comment]
There is a reason for the madness of CSS naming (If it's good reasoning is a different question). But it is a mix of namespacing to avoid collision and BEM naming conventions to avoid specificity issues.

With global CSS and without something like CSS modules we can have issues with collisions. Especially with general things like "button". Yes, that shouldn't be done, but typically it can be and it's hard to detect (and often cause regression issues).
So to avoid this I started namespacing things, especially for UI components. I see I forgot to amend the first components I started with, though.

For using BEM (block, element, modifier) it is to avoid fighting with specificity and to semantically group styling together. I typically avoid using nested concatenative CSS with SCSS/Less as it makes it harder to search for CSS and lookup from code to CSS.

TL;DR, the structure here is:

```
.namespace-separated-by-kebabs-moduleNameAsPartOfNamespace__blockElement__subBlock--modifierName {}
```

File: src/elm/Page/Shop.elm (hunk 7 above)
Comment (mikaelbr) [reply src/elm/Page/Shop.elm:12]: 😱  VS Code sometimes do autoimporting

File: src/static/styles/main.scss (hunk 2 above)
Comment (mjenssen) [reply src/static/styles/main.scss:27]: It says 11px in computed in my Firefox, but I see it tries to use `1rem` in the rules tab.

Comment (mikaelbr) [reply src/static/styles/main.scss:27]: Also. I've removed some of the default styles for button/inputs which cause some issues elsewhere. But going forward using UI Components this will be resolved.

File: src/elm/Ui/Input.elm (hunk 1 above)
Comment (mikaelbr) [reply src/elm/Ui/Input.elm:15]: https://github.com/AtB-AS/webshop/issues/42

File: src/static/styles/main.scss (hunk 2 above)
Comment (mikaelbr) [reply src/static/styles/main.scss:27]: > It says 11px in computed in my Firefox, but I see it tries to use 1rem in the rules tab.

Yep, but I pushed explicit 16px now, which should fix it for now. Mainly this is for iOS phone zoom though. The design is going to get somewhat in flux until we refactor code to use UI components. I'm gradually removing default styling on things to avoid having to reset it and to be more interoperable/pluggable within different contexts

**Comment by mjenssen** [comment]
> TL;DR, the structure here is:
> 
> ```css
> .namespace-separated-by-kebabs-moduleNameAsPartOfNamespace__blockElement__subBlock--modifierName {}
> ```

Yeah, I knew it was BEM, but I had forgotten just how crazy of a mess that thing is in its full glory. Thanks for clarifying!

**Review by mjenssen — APPROVED** [review]
LGTM
